src/electron/routing_service.ts

```diff
@@ -251,52 +291,68 @@ function installWindowsRoutingServices(): Promise<void> {
+  if (!isLinux) {
+    throw new Error('please call this method in Linux');
   }
-  console.log(`all service installation files copied to ${tmp} successfully`);
 
-  await executeCommandAsRoot(path.join(tmp, LINUX_INSTALLER_FILENAME));
+  console.log('installing Outline service...');
+  const SERVICE_SOURCE_FOLDER = path.resolve(getAppPath(), 'tools/outline_proxy_controller/dist');
+  const SERVICE_BINARY = 'OutlineProxyController';
+  const SERVICE_BINARY_TARGET_FOLDER = '/usr/local/sbin';
+  const SERVICE_DEFINITION = 'outline_proxy_controller.service';
+  const SERVICE_DEFINITION_TARGET_FOLDER = '/etc/systemd/system';
+
+  const binSource = path.join(SERVICE_SOURCE_FOLDER, SERVICE_BINARY);
+  const binTarget = path.join(SERVICE_BINARY_TARGET_FOLDER, SERVICE_BINARY);
+  console.log(`copying Outline service binary from "${binSource}" to "${binTarget}"...`);
+  await fsextra.copy(binSource, binTarget, {overwrite: true});
+  await fsextra.chmod(binTarget, 0o755);
+  console.info('successfully installed Outline service binary');
+
+  const svcSource = path.join(SERVICE_SOURCE_FOLDER, SERVICE_DEFINITION);
+  const svcTarget = path.join(SERVICE_DEFINITION_TARGET_FOLDER, SERVICE_DEFINITION);
+  console.log(`copying Outline service definition from "${svcSource}" to "${svcTarget}"...`);
+  await fsextra.copy(svcSource, svcTarget, {overwrite: true});
+  console.info('successfully installed Outline service definition');
+
+  console.log('registering Outline service to systemctl...');
+  await executeShellCommand('systemctl daemon-reload');
+  await executeShellCommand(`systemctl enable ${SERVICE_DEFINITION}`);
+  await executeShellCommand(`systemctl restart ${SERVICE_DEFINITION}`);
+  console.info('Outline service successfully registered');
+
+  // Because the .service file specifies Type=simple, the installation script exits immediately.
+  // Sleep for a couple of seconds before exiting.
+  await new Promise(r => setTimeout(r, 2000));
+  console.info('Outline service successfully installed');
 }
 
 export async function installRoutingServices(): Promise<void> {
   console.info('installing outline routing service...');
   if (isWindows) {
-    await installWindowsRoutingServices();
+    await doInstallWindowsRoutingServices();
   } else if (isLinux) {
-    await installLinuxRoutingServices();
+    // Run `sudo ./Outline-Client.AppImage --install` to install the service
+    //   * `--headless` is required by chromium:
+    //       'ERROR:ozone_platform_x11.cc(247)] Missing X server or $DISPLAY'
+    //   * `$DISPLAY` and `xhost +si:localuser:root` is required by electron:
+    //       'Gtk-WARNING **: 18:24:17.308: cannot open display: '
+    //   * `--no-sandbox` is required for root due to chrome issue: https://crbug.com/638180
+    //   * gpu related options are used to prevent any potential GPU permission issues
+    const oldRootXServerAccess = await configureRootXServerAccess(true);
```

Comment:
This (`xhost +`) and the arguments such as `--headless` below resolve permission issues for root to access GUI apps. Chromium supports headless (non-GUI) but electron does not.

I feel creating a pure CLI app would be ideal for this case, for example, we can reuse the `OutlineProxyController` binary and implement `./OutlineProxyController --install`. To prevent command injections, we can:

```typescript
const tmp = await fsextra.mkdtemp('/tmp/');
const svcBinary = `${tmp}/OutlineProxyController `;
await fsextra.copy('.../.mountXXXX/tools/outline_proxy_controller/dist/OutlineProxyController', svcBinary, {overwrite: true});
await executeCommandAsRoot(
  '{ ' +
    `chattr +i '${svcBinary}' && ` +
    `echo '${chksum}  ${svcBinary}' | shasum -c && ` +
    `'${svcBinary}' --install` +
  `}; chattr -i '${svcBinary}'`);
```

Reply:
I like that idea, looks a lot cleaner. Do you wanna implement that?

I think you can actually call `chattr -i` as the user. Once you do that, only root can make it mutable (see https://man7.org/linux/man-pages/man2/ioctl_iflags.2.html). Then you can also check the hash without using root.

Reply:
Great! Let me try that. Also I prefer this idea because as we talked before, there is a similar security issue in `sudo-prompt` package for Windows, we can [force the `exe` running under admin permission by turning on the `/MANIFESTUAC:requireAdministrator` link option](https://docs.microsoft.com/en-us/cpp/build/reference/manifestuac-embeds-uac-information-in-manifest?view=msvc-170), so we can achieve the similar behavior by calling `OutlineService.exe --install` without using `sudo-prompt` package.